golem-compilation-server/src/service/upload_worker.rs

```diff
@@ -0,0 +1,72 @@
+use std::sync::Arc;
+
+use futures::stream::StreamExt;
+use golem_common::config::RetryConfig;
+use golem_worker_executor_base::services::compiled_template::CompiledTemplateService;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+use tokio_util::sync::CancellationToken;
+
+use crate::model::*;
+
+// Worker that uploads compiled templates to the cloud.
+#[derive(Clone)]
+pub struct UploadWorker {
+    retry_config: RetryConfig,
+    compiled_template_service: Arc<dyn CompiledTemplateService + Send + Sync>,
+}
+
+impl UploadWorker {
+    pub fn start(
+        retry_config: RetryConfig,
+        num_workers: usize,
+        compiled_template_service: Arc<dyn CompiledTemplateService + Send + Sync>,
+        recv: mpsc::Receiver<CompiledTemplate>,
+        cancellation: CancellationToken,
+    ) {
+        let worker = Self {
+            retry_config,
+            compiled_template_service,
+        };
+
+        let mut stream = ReceiverStream::new(recv)
```

Comment:
Multiple Upload "workers" uploading compiled files with buffered parallelism

Reply:
@vigoo

Reply:
I don't think it matters much to parallelize these uploads - I expect the compilation itself takes orders of magnitude more time, so this will likely never really parallelize anything. We can keep it simple and just do it one by one

Reply:
yeah you're completely right. I didn't account for compilation happening slower, so there's no need for parallel uploads. great catch.